src/tss2-tcti/tcti-tbs.c

```diff
@@ -0,0 +1,291 @@
+    tcti_tbs->resSize = 0;
+    return rc;
+}
+
+void
+tcti_tbs_finalize (
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs = tcti_tbs_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_tbs_down_cast (tcti_tbs);
+    TBS_RESULT tbs_rc;
+
+    if (tcti_tbs == NULL) {
+        return;
+    }
+
+    free (tcti_tbs->resultBuffer);
+
+    tbs_rc = Tbsip_Context_Close (tcti_tbs->hContext);
+    if (tbs_rc != TBS_SUCCESS) {
+        LOG_WARNING ("Failed to close context with TBS error: 0x%x", tbs_rc);
+    }
+
+    tcti_common->state = TCTI_STATE_FINAL;
+}
+
+TSS2_RC
+tcti_tbs_cancel (
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    TBS_RESULT tbs_rc;
+    TSS2_RC rc = TSS2_RC_SUCCESS;
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs = tcti_tbs_context_cast (tctiContext);
+
+    tbs_rc = Tbsip_Cancel_Commands (tcti_tbs->hContext);
+    if (tbs_rc != TBS_SUCCESS) {
+        LOG_WARNING ("Failed to cancel commands with TBS error: 0x%x", tbs_rc);
+        rc = TSS2_TCTI_RC_GENERAL_FAILURE;
+    }
+
+    return rc;
+}
+
+TSS2_RC
+tcti_tbs_get_poll_handles (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    TSS2_TCTI_POLL_HANDLE *handles,
+    size_t *num_handles)
+{
+    /* TBS doesn't support polling. */
+    (void)(tctiContext);
+    (void)(handles);
+    (void)(num_handles);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+TSS2_RC
+tcti_tbs_set_locality (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    uint8_t locality)
+{
+    /*
+     * TBS currently only supports locality 0
+     */
+    (void)(tctiContext);
+    (void)(locality);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+TSS2_RC
+Tss2_Tcti_Tbs_Init (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t *size,
+    const char *conf)
+{
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs;
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common;
+    TBS_RESULT tbs_rc;
+    TBS_CONTEXT_PARAMS2 params;
+    TPM_DEVICE_INFO info;
+
+    if (tctiContext == NULL) {
+        if (size == NULL) {
+            return TSS2_TCTI_RC_BAD_VALUE;
+        }
+        *size = sizeof (TSS2_TCTI_TBS_CONTEXT);
+        return TSS2_RC_SUCCESS;
+    }
+
+    /* Init TCTI context */
+    TSS2_TCTI_MAGIC (tctiContext) = TCTI_TBS_MAGIC;
+    TSS2_TCTI_VERSION (tctiContext) = TCTI_VERSION;
+    TSS2_TCTI_TRANSMIT (tctiContext) = tcti_tbs_transmit;
+    TSS2_TCTI_RECEIVE (tctiContext) = tcti_tbs_receive;
+    TSS2_TCTI_FINALIZE (tctiContext) = tcti_tbs_finalize;
+    TSS2_TCTI_CANCEL (tctiContext) = tcti_tbs_cancel;
+    TSS2_TCTI_GET_POLL_HANDLES (tctiContext) = tcti_tbs_get_poll_handles;
+    TSS2_TCTI_SET_LOCALITY (tctiContext) = tcti_tbs_set_locality;
+    TSS2_TCTI_MAKE_STICKY (tctiContext) = tcti_make_sticky_not_implemented;
+    tcti_tbs = tcti_tbs_context_cast (tctiContext);
+    tcti_common = tcti_tbs_down_cast (tcti_tbs);
+    tcti_common->state = TCTI_STATE_TRANSMIT;
+
+    memset (&tcti_common->header, 0, sizeof (tcti_common->header));
+    tcti_common->locality = 0;
+
+    params.includeTpm20 = 1;
```

Comment:
Does this leave the `includeTpm12` field uninitialized? I'm reading through the docs for the `Tbsi_Context_Create` function and if both are set (or one uninitialized and set unintentionally) then the create function may return a context that's communicating with a 1.2 TPM. You check below using the `Tbsi_GetDeviceInfo` function so the code will catch the error but if we can only communicate with TPM2 devices then initializing this field to `0` (possibly where the structure is declared) makes sense. Can't hurt to check again via `GetDeviceInfo` as well if we don't trust / want to double check the create function.

Reply:
This should be fixed now, `includeTpm12` is now initialized to `0` on the line below this. I plan on keeping the `GetDeviceInfo` check just in case though.